tests/testing_utils.py

```diff
@@ -61,6 +67,100 @@ def run_command_with_inputs(command_list: List[str], inputs: List[str], timeout=
     return run_command_with_input(command_list, ("\n".join(inputs) + "\n").encode(), timeout)
 
 
+def start_persistent_process(
+    command_list: List[str],
+    cwd: Optional[str] = None,
+) -> Popen:
+    """Start a process with parameters that are suitable for persistent execution."""
+    return Popen(
+        command_list,
+        stdout=PIPE,
+        stderr=subprocess.STDOUT,
+        stdin=PIPE,
+        encoding="utf-8",
+        bufsize=1,
+        cwd=cwd,
+    )
+
+
+def kill_process(process: Popen) -> None:
+    """Kills a process and it's children.
+    This loop ensures orphaned children are killed as well.
+    https://psutil.readthedocs.io/en/latest/#kill-process-tree
+    Raises ValueError if some processes are alive"""
+    root_process = psutil.Process(process.pid)
+    all_processes = root_process.children(recursive=True)
+    all_processes.append(root_process)
+    for process in all_processes:
+        try:
+            process.kill()
+        except psutil.NoSuchProcess:
+            pass
+    _, alive = psutil.wait_procs(all_processes, timeout=10)
+    if alive:
+        raise ValueError(f"Processes: {alive} are still alive.")
+
+
+def read_until_string(process: Popen, expected_output: str, timeout: int = 5) -> None:
+    """Read output from process until a line equals to expected_output has shown up or reaching timeout.
+    Throws TimeoutError if times out
+    """
+
+    def _compare_output(output, outputs):
+        return output == expected_output
+
+    try:
+        read_until(process, _compare_output, timeout)
+    except TimeoutError as ex:
+        expected_output_bytes = expected_output.encode("utf-8")
+        raise TimeoutError(
+            f"Did not get expected output after {timeout} seconds. Expected output: {expected_output_bytes!r}"
+        ) from ex
+
+
+def read_until(process: Popen, callback: Callable[[str, List[str]], None], timeout: int = 5):
+    """Read output from process until callback returns True or timeout is reached
+
+    Parameters
+    ----------
+    process : Popen
+    callback : Callable[[str, List[str]], None]
+        Call when a new line is read from the process.
+    timeout : int, optional
+        By default 5
+
+    Raises
+    ------
+    TimeoutError
+        Raises when timeout is reached
+    """
+    result_queue: Queue = Queue()
+
+    def _read_output():
+        try:
+            outputs = list()
+            for output in process.stdout:
+                outputs.append(output)
+                LOG.info(output.encode("utf-8"))
+                if callback(output, outputs):
+                    result_queue.put(True)
+                    return
+        except Exception as ex:
+            result_queue.put(ex)
+
+    reading_thread = Thread(target=_read_output, daemon=True)
+    reading_thread.start()
+    reading_thread.join(timeout=timeout)
+    if reading_thread.isAlive():
```

Comment:
Should this method kill the process if it reaches after timeout?

Reply:
I want to leave that decision up for each individual test case instead. There might be valid use cases where we want to read another line even if the current line timed out.
For the watch integration test, I put the code for killing the process in teardown instead.